languages/rust/exercises/concept/tuples/.docs/introduction.md

```diff
@@ -0,0 +1,83 @@
+# Tuples
+
+Tuples are a lightweight way to group a bounded set of heterogenous data together. A tuple doesn't have
+a particular name; naming a data structure turns it into a `struct`. A tuple's fields don't have names;
+they are accessed by means of destructuring or by position.
+
+## Syntax
+
+Tuples are a list of fields surrounded by parentheses:
+
+- `()` is the `unit type`, which is a [special case](https://doc.rust-lang.org/std/primitive.unit.html)
+- **(** _expression_ **,)** is a single-element tuple expression
+- **(** _Type_ **,)** is a single-element tuple type
+- **(** _expression_ **,** ... **)** is a general tuple expression
+- **(** _Type_ **,** ... **)** is a general tuple type
+
+### Creation
+
+Tuples are always created with a tuple expression:
+
+```rust
+// pointless but legal
+let unit = ();
+// single element
+let single_element = ("note the comma",);
+// two element
+let two_element = (123, "elements can be of differing types");
+```
+
+Tuples can have an arbitrary number of elements.
+
+### Access by destructuring
+
+It is possible to access the elements of a tuple by destruturing. This just means assigning variable
+names to the individual elements of the tuple, consuming it.
+
+```rust
+let (elem1, _elem2) = two_element;
+assert_eq!(elem1, 123);
+```
+
+### Access by position
+
+It is also possible to access the elements of a tuple by numeric positional index. Indexing, as always,
+begins at 0.
+
+```rust
+let notation = single_element.0;
+assert_eq!(notation, "note the comma");
+```
+
+## Tuple Structs
+
+You will also be asked to work with tuple structs. Like normal structs, these are named types; unlike
+normal structs, they have anonymous fields. Their syntax is very similar to normal tuple syntax. It is
+legal to use both destructuring and positional access.
+
+```rust
+struct TupleStruct(u8, i32);
+let my_tuple_struct = TupleStruct(123, -321);
+let borrowed_neg = &my_tuple_struct.1;
+let TupleStruct(borrowed_byte, _) = &my_tuple_struct;
+assert_eq!(borrowed_neg, &-321);
+assert_eq!(borrowed_byte, &123);
```

Comment:
Do students know what the `&` operator is and what borrowing means when they'll work on this exercise?

Reply:
Probably not, at this point. That's ok, though: this is an example. Students are not required to use or understand borrowing at any point in this exercise.

Borrowing is pervasive in Rust; it's not practical or desirable to shield students from all mention of it until the borrowing exercise. Instead, we can use natural examples of it to help them get familiar with the way it looks before we actually teach it.

That said, if you feel strongly, I'm willing to change this example.

Reply:
> That said, if you feel strongly, I'm willing to change this example.

It's not that I personally feel strongly about it, it's just how we designed v3 to work :)

Reply:
@coriolinus I'd like to expand a little on this for clarity.

I don't philosophically disagree with your opinion that it is helpful to expose students to things. For me individually, I quite like that approach when I'm learning. However, I have learnt that this can be extremely offputting to people, as they feel the need to be able to understand everything they read before they can move on. So showing them the `&` syntax means that they feel the need to go and learn the `&` syntax and everything about it before they solve the exercise. The same is true of introducing algorithms - even when we say "don't worry about this" people still feel the need to go and read everything about an algorithm if it's part of an exercise. 

I've spoken to lots of people about this (including both junior-level devs and extremely-senior devs) and this is an extremely common mentality. Maybe the majority in fact.

This is why we've made the decision that the only things that we show to students should be things that we've explained and that they need to use to solve an exercise. Extra examples (such as the borrowing syntax) can appear in the concept's `about.md` because students don't have to read that in order to solve the exercise so it's not part of their "flow". 

The same is true of the iterators discussion above - regardless of us saying "don't worry about this", people will, and we'll have a big drop-off at this point because of it. But I want to think more about that example before replying to try and come up with a solution rather than just an objection :)

Reply:
OK, I've changed it in 9e386eb5. That said, we haven't eliminated borrowing or the `&` sigil entirely from the exercise: for example, in

```rust
pub fn manhattan(&self) -> i16 { ... }
```

That function takes `&self` instead of `self` because it doesn't consume `self`; you can call it more than once on the same instance. That's the normal and expected behavior, but it needs to remain as syntactic noise for the students at this point in the concept graph. 

In principle, we could write that as a consuming function. That would teach bad habits, though, because it would have the wrong semantics. I feel strongly that a little bit of syntactic noise that students need to overlook is a smaller price to pay than teaching bad habits which will come back to bite the students later.

Reply:
Thanks. 

If the code is in the provided stub then I'm ok with it containing things that a student doesn't need to know. I see that as a "necessary evil" so as not to teach wrong-habits, exactly as you say.